(fix) Segmentation fault when using a compound type.

jwsblokland opened the pull request:
In the case when a compounded attribute is written to dataset
followed by writing the data with a data transform function
to the dataset will result in a segmentation fault. It turns out
the data is classified as compounded while it is not. Now, the
state is always reset first to not compounded followed by the
existing check if the variable is compounded.

An example of the problem and to show it is solved by this pull request: [compound.txt](https://github.com/HDFGroup/hdf5/files/5614090/compound.txt)

This pull request solves the items SUPPORT-857 and HDFFV-11050

Commit:
(fix) Segmentation fault when using a compound type.

In the case when a compounded attribute is written to dataset
followed by writing the data with a data transform function
to the dataset will result in a segmentation fault. It turns out
the data is classified as compounded while it is not. Now, the
state is always reset first to not compounded followed by the
existing check if the variable is compounded.

derobins commented on src/H5T.c:
```diff
@@ -5217,7 +5217,12 @@ H5T__path_find_real(const H5T_t *src, const H5T_t *dst, const char *name, H5T_co
-     * for the optimization of data reading (in H5Dio.c). */
+     * for the optimization of data reading (in H5Dio.c).
+     *
+     * Jan-Willem Blokland, 11 March 2020
```
Please remove this line. It's going to be really distracting if the code fills up with line-by-line credit for edits.

Also, feel free to remove Quincey's similar edit, above :)

jwsblokland replied:
Good point. I will remove it.

derobins commented:
This should get a JIRA issue and needs a test demonstrating the problem and ensuring it's fixed. Do you have a minimal test program that fails? If you can send me a test program that confirms the bug, I'll create a JIRA issue. I can also point you to where you should add a test and explain our test harness a bit, but that will have to wait until tomorrow morning.

Commit:
(fix) Removed undesired comment lines.

jwsblokland commented:
The included example compound.txt is an minimal test program. It is c source file but github does not allow me to upload a file with the extension .c. 

Yes if you can explain your test setup, I will add a test for this case.

Furthermore, for your convenience I added the existing JIRA issue numbers SUPPORT-857 and HDFFV-11050 in which the problem is described.

qkoziol commented:
> The included example compound.txt is an minimal test program. It is c source file but github does not allow me to upload a file with the extension .c.
> 
> Yes if you can explain your test setup, I will add a test for this case.
> 
> Furthermore, for your convenience I added the existing JIRA issue numbers SUPPORT-857 and HDFFV-11050 in which the problem is described.

Yes, a test would be needed for this.

Commit:
(fix) Segmentation fault when using a compound type: added test.

Commit:
(fix) Added the missing cmpd_transform.c file to MANIFEST.

jwsblokland commented:
Added a new test to the test suite to demonstrate the segmentation fault is resolved.

qkoziol reviewed (changes requested):
Looks good, but can you please add the new test to the autotools framework also?

lrknox commented on test/cmpd_dtransform.c:
```diff
@@ -0,0 +1,139 @@
+ * is linked from the top-level documents page.  It can also be found at     *
+ * http://hdfgroup.org/HDF5/doc/Copyright.html.  If you do not have          *
+ * access to either file, you may request a copy from help@hdfgroup.org.     *
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
```
The HDF Group copyright header has changed, and the line "* Copyright by the Board of Trustees of the University of Illinois." applies only to files created before 2007.
The text in the paragraph has changed and the link that is now for the COPYING file also needs updating in all of the HDF5 source files, which will be happening within a few days, so it isn't necessary to make changes for this pull request, especiallly since pasting it here made format changes.  Once that update is done, this header will be replaced with the following:
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *  
 * Copyright by The HDF Group.                                                                 *
 * All rights reserved.                                                                                   *
 
 * This file is part of HDF5.  The full HDF5 copyright notice, including       *
 * terms governing use, modification, and redistribution, is contained in  *
 * the COPYING file, which can be found at the root of the source code   *
 * distribution tree, or in https://www.hdfgroup.org/licenses.                    *
 * If you do not have access to either file, you may request a copy from   *
 * help@hdfgroup.org.                                                                                 *

jwsblokland replied:
I have updated the source header to the latest version.

lrknox replied:
Thanks!

Commit:
(fix) cmpd_dtransform test: autotools and source header.

Added the cmp_dtransform test to the autotools configuration and
updated the HDF Group copyright header.

jwsblokland commented:
The new test has been added to the autotools.

qkoziol commented:
> The new test has been added to the autotools.

Super, thanks!
